### PR TITLE
All of any of one of

### PR DESCRIPTION
**Pull request by adamcorina**

### CODE REVIEW

File: src/schema-form.tsx [hunk 1]
```diff
@@ -62,21 +56,9 @@ export const SchemaForm = ({
...
-                        const: ifEntry[1].const,
-                        ...(schema.then ? { then: schema.then } : {}),
-                        ...(schema.else ? { else: schema.else } : {}),
-                        lastCondition: ''
-                    }
-                }
-            }
-        }
-        return {}
+    const [conditionals] = useState<{ data: Conditional[]; lastEvaluations: string }>(() => {
```
Comment (cstefanache) [inline src/schema-form.tsx:59]: Interface for { data: Conditional[]; lastEvaluations: string } maybe?

File: src/schema-form.tsx [hunk 2]
```diff
@@ -93,31 +75,16 @@ export const SchemaForm = ({
...
-                    }
-                    addProperties(currentSchema[key], newProperties[key])
-                } else {
-                    Object.assign(currentSchema, { [key]: newProperties[key] })
-                }
+    const removeProperties = (currentSchema: any, baseSchema: any, nestedPath: string) => {
+        if (isObject(currentSchema) && isObject(baseSchema)) {
+            if (currentSchema['$ref'] !== undefined) {
+                const def = currentSchema['$ref'].substr(currentSchema['$ref'].lastIndexOf('/') + 1)
+                addProperties(currentSchema, root!.definitions[def])
```
Comment (cstefanache) [inline src/schema-form.tsx:82]: what happens if definitions is not defined on root? (I assume that addProperties will be executed adding another call on the stack) - if yes can we do a check before executing `addProperties`

File: src/schema-form.tsx [hunk 3]
```diff
@@ -171,17 +179,17 @@ export const SchemaForm = ({
...
     }
 
     const handleSubmit = () => {
         const result = instance!.validate(obj.data)
         const errors: ajv.ErrorObject[] = result || !instance!.validator.errors ? [] : instance!.validator.errors
-        errors.forEach((err) => {
-            if (err.params && (err.params as RequiredParams).missingProperty) {
+
+        errors.forEach((err, index, object) => {
+            if (err.keyword === 'if') {
```
Comment (cstefanache) [inline src/schema-form.tsx:190]: we should start a constants file to avoid string usage

File: src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+export const addProperties = (currentObject: any, newProperties: any): any => {
+    if (isObject(currentObject) && isObject(newProperties)) {
+        for (const key in newProperties) {
+            if (isObject(newProperties[key])) {
+                if (!currentObject[key]) {
+                    Object.assign(currentObject, { [key]: {} })
+                }
+                addProperties(currentObject[key], newProperties[key])
+            } else {
+                if(key === 'required' && Array.isArray(newProperties[key]) && Array.isArray(currentObject[key])){
```
Comment (cstefanache) [inline src/utils.ts:25]: hmmm ... is there a chance that the Schema could be defined with uppercase key names? "REQUIRED": "whatever" ? If yes I hate the one who does this!

Comment (cstefanache) [reply src/utils.ts:25]: hate ones that write using uppercase

File: src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    ifEntries = ifEntries.concat(getMultipleConditionals(schema))
+    return ifEntries.map((ifEntry) => getCompiledConditional(ifEntry))
+}
+
+//given a schema, the function will return a list with all the paths of the schema and the value of the leaf node
+//keywords such as properties and const are not included
+//usage: extract all the conditions specified inside an if statement
+const getPropertyPaths = (schema: any, path = ''): { '0': string; '1': any }[] => {
+    if (isObject(schema)) {
+        return Array.prototype.concat.apply(
```
Comment (cstefanache) [inline src/utils.ts:53]: `return [...Object.keys(schema).map((key) =>
                getPropertyPaths(schema[key], key === 'properties' || key === 'const' ? path : path + '.' + key)
            )]`
?

Comment (adamcorina) [reply src/utils.ts:53]: I want to get the data paths of all properties in an if statement; the function has to ignore the properties layer of the schema (it does not exist on the data object), as well as the const keyword; they will not be appendend to the path;
e.g. the following schema given as parameter 

{
          "properties": {
                "second": {
                    "const": true
                 }
           }
 }

will return

[{'second': true}]

File: src/utils.ts (hunk 4 above)
Comment (adamcorina) [reply src/utils.ts:25]: this case will be ignored for now :)

**Pull request merged**